web_api.md

```diff
@@ -245,8 +245,47 @@ Instead of `path_to_mets`, one can also use `workspace_id` to specify a workspac
 from the Workspace Server. In case `result_queue_name` property is presented, the result of the processing will be
```

Comment:
Sorry, but I still don't understand. So the Processing Server (formerly Broker) receives some `POST /processor/{exe}` or `POST /workflow/{wid}` with parameters, respectively. It generates a job ID. It then divides the request up into a chain of processor jobs, creating a message to the queue for each one. Each queue is named by the processor executable.

- Is `result_queue_name` the same as the `appropriate queue` mentioned in the `Processing` paragraph of the architecture?
- How do the workers know when they have to skip some message because its prerequisite message has not finished, yet?

Reply:
> Is `result_queue_name` the same as the `appropriate queue` mentioned in the Processing paragraph of the architecture?

No, `appropriate queue` means the processing queues, which have the same name as the corresponding processor. For example, `ocrd-dummy` processors always listen to `ocrd-dummy` queue.

> How do the workers know when they have to skip some message because its prerequisite message has not finished, yet?

It doesn't know that. A worker is completely stateless. It only takes a message from a queue and start processing. It's the job of some kind of orchestrator above to push messages to queues in the right order.

Reply:
Understood. Your additional terminology on job queue and result queue is also helpful, thanks.

But I still don't get the picture. So if there are multiple job queues (one for each available processor), but the workers are stateless, then how does the synchronisation between producers and consumers happen?

Reply:
> So if there are multiple job queues (one for each available processor)

No, one queue for each type of processor. For example, 2 `ocrd-dummy` processors will listen to the same `ocrd-dummy` queue.

> how does the synchronisation between producers and consumers happen?

An example flow looks like this:

1. The producer, which in our case is the Processing Server, receives a `POST` request to the `/processor/ocrd-dummy` endpoint with this body:

```yml
path_to_mets: /path/to/mets.xml
input_file_grps:
  - OCR-D-DEFAULT
output_file_grps:
  - OCR-D-BIN
page_id: PHYS_001,PHYS_002
parameters:
  params_1: 1
  params_2: 2

result_queue_name: ocrd-dummy-result
callback_url: https://my.domain.com/callback
```

2. It creates a processing message based on this content and publish it to the `ocrd-dummy` queue.
3. An `ocrd-dummy` worker, which is listening to that queue, receives the message and starts processing.
4. When the processing finished, the worker saves the job status to the database, publishes a result message to the `result_queue_name` and also `POST`s the result message to the `callback_url`.

I hope that answers your question.

Reply:
> > So if there are multiple job queues (one for each available processor)
> 
> No, one queue for each type of processor. For example, 2 `ocrd-dummy` processors will listen to the same `ocrd-dummy` queue.

That's what I meant. (Note I did not say _worker_ but _processor.)

> > how does the synchronisation between producers and consumers happen?
> 
> An example flow looks like this:

This example is too trivial – I already understood that much. What I meant by synchronisation between producers and consumers is that in every non-trivial workflow (more than 1 step), processors run one after another (for the same workspace), and each step (producing processor)'s output is the next step (consuming processor)'s input. IOW the consumer must not run before the producer.

Now, if the workers only get to see their own respective queue, and the steps are scattered across many queues – how is the workflow queued? (Or is there some runtime instance – like a thread for each workflow – that tracks progress, pushing new process messages to the job queues after receiving the previous result messages?)

Reply:
Ah, I see what you meant. In our current plan:

1. All processing workers are only consumers. It means that they won't publish any processing messages.
2. We need an orchestrator (Nextflow in our case) to execute a workflow. The orchestrator listens to the result queues and send the next `POST /processing/{exe}` request to the Processing Server to trigger the next processor when the previous one is finished.

For the next step, I'm thinking it might make sense to have a simple workflow definition file (YAML), in case Nextflow is overkill for any use cases. By reading that file, a processing worker knows what to do and which processor it has to trigger next when it finished. So in this case, a processing worker will also produce processing messages. But this is only the plan in my head. We need to finish what is defined in the current spec first.

Reply:
Ah, of course, now I see (the wood from the trees). 

So perhaps the spec should also contain an example for the workflow server (detailling how the workflow gets decomposed into jobs by processor or even by page, and how the job/result queue communication then looks like)?

> For the next step, I'm thinking it might make sense to have a simple workflow definition file (YAML), in case Nextflow is overkill for any use cases. By reading that file, a processing worker knows what to do and which processor it has to trigger next when it finished. So in this case, a processing worker will also produce processing messages. But this is only the plan in my head. We need to finish what is defined in the current spec first.

Interesting idea.

Reply:
As discussed, an example Nextflow script will be provided later.